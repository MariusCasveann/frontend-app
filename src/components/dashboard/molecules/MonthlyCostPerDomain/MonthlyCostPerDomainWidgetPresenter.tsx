import React from 'react';
import * as chartjs from 'chart.js';
import { Bar } from 'react-chartjs-2';
import LoadingOrErrorPresenter from '../../../common/organisms/LoadingOrError/LoadingOrErrorPresenter';
import { hasChartData } from '../../helpers/checkChartData';
import './MonthlyCostPerDomain.css';
import { MESSAGE_WHEN_NO_DATA } from '../../../../utils/constants';

interface MonthlyCostPerDomainProps {
    chartData?: chartjs.ChartData;
    error: string;
    loading: boolean;
}

export default ({ chartData, error, loading }: MonthlyCostPerDomainProps) => {
    if (loading || error) {
        return <LoadingOrErrorPresenter error={error} loading={loading} />;
    }
    if (hasChartData(chartData)) {
        return (
            <div className="chart">
                {chartData && (
                    <Bar
                        data={chartData}
                        options={{
                            scales: {
                                xAxes: [
                                    {
                                        barPercentage: 0.3
                                    }
                                ]
                            },
                            responsive: true,
                            maintainAspectRatio: false,
                            legend: {
                                display: false
                            }
                        }}
                    />
                )}
            </div>
        );
    }
    return <p>{MESSAGE_WHEN_NO_DATA}</p>;
};
