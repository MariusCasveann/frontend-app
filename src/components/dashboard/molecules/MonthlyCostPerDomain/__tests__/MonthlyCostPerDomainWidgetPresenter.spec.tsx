import * as React from 'react';
import renderer from 'react-test-renderer';
jest.mock('react-chartjs-2', () => ({ Bar: 'Bar' }));
jest.mock('../../../../common/organisms/LoadingOrError/LoadingOrErrorPresenter', () => 'LoadingOrErrorPresenter');
import MonthlyCostPerDomainWidgetPresenter from '../MonthlyCostPerDomainWidgetPresenter';

const mockData = {
    labels: ['cloud-ccr', 'pref-center', 'promotions', 'rea'],
    datasets: [
        {
            data: [2111.74, 159.63, 802.89, 222.99],
            backgroundColor: '#859ec4ÃŸ',
            label: 'Monthly cost'
        }
    ]
};

const mockEmptyData = {
    labels: [],
    datasets: [
        {
            data: [],
            backgroundColor: '#859ec4',
            label: 'Monthly cost'
        }
    ]
};

describe('MonthlyCostPerDomainWidgetPresenter', () => {
    describe('when there are chart data', () => {
        it('renders a Bar widget', () => {
            const wrapper = renderer
                .create(<MonthlyCostPerDomainWidgetPresenter error={''} loading={false} chartData={mockData} />)
                .toJSON();
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('when there are no chart data', () => {
        it('doesnt render a Bar widget', () => {
            const wrapper = renderer
                .create(<MonthlyCostPerDomainWidgetPresenter error={''} loading={false} chartData={mockEmptyData} />)
                .toJSON();
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('when chart data is undefined', () => {
        it('renders loading', () => {
            const wrapper = renderer
                .create(<MonthlyCostPerDomainWidgetPresenter error={''} loading={true} chartData={undefined} />)
                .toJSON();
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('when chart data is undefined', () => {
        it('renders an error', () => {
            const wrapper = renderer
                .create(<MonthlyCostPerDomainWidgetPresenter error={'Error'} loading={false} chartData={undefined} />)
                .toJSON();
            expect(wrapper).toMatchSnapshot();
        });
    });
});
