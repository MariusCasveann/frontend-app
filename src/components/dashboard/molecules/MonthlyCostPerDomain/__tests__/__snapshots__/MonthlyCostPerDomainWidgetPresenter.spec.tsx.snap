// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonthlyCostPerDomainWidgetPresenter when chart data is undefined renders an error 1`] = `
<LoadingOrErrorPresenter
  error="Error"
  loading={false}
/>
`;

exports[`MonthlyCostPerDomainWidgetPresenter when chart data is undefined renders loading 1`] = `
<LoadingOrErrorPresenter
  error=""
  loading={true}
/>
`;

exports[`MonthlyCostPerDomainWidgetPresenter when there are chart data renders a Bar widget 1`] = `
<div
  className="chart"
>
  <Bar
    data={
      Object {
        "datasets": Array [
          Object {
            "backgroundColor": "#859ec4ÃŸ",
            "data": Array [
              2111.74,
              159.63,
              802.89,
              222.99,
            ],
            "label": "Monthly cost",
          },
        ],
        "labels": Array [
          "cloud-ccr",
          "pref-center",
          "promotions",
          "rea",
        ],
      }
    }
    options={
      Object {
        "legend": Object {
          "display": false,
        },
        "maintainAspectRatio": false,
        "responsive": true,
        "scales": Object {
          "xAxes": Array [
            Object {
              "barPercentage": 0.3,
            },
          ],
        },
      }
    }
  />
</div>
`;

exports[`MonthlyCostPerDomainWidgetPresenter when there are no chart data doesnt render a Bar widget 1`] = `
<p>
  No data available, please select another time and/or domain/product.
</p>
`;
